lib/membrane_opus/encoder.ex

```diff
@@ -0,0 +1,200 @@
+defmodule Membrane.Opus.Encoder do
+  @moduledoc """
+  This element performs encoding of Opus audio into a raw stream. You'll need to parse the stream and then package it into a container in order to use it.
+  """
+
+  use Membrane.Filter
+  use Bunch.Typespec
+
+  alias __MODULE__.Native
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw
+  alias Membrane.Caps.Matcher
+
+  @list_type allowed_channels :: [1, 2]
+
+  @list_type allowed_applications :: [:voip, :audio, :low_delay]
+  @default_application :audio
+
+  @list_type allowed_sample_rates :: [8000, 12_000, 16_000, 24_000, 48_000]
+
+  @supported_input {Raw,
+                    format: :s16le,
+                    channels: Matcher.one_of(@allowed_channels),
+                    sample_rate: Matcher.one_of(@allowed_sample_rates)}
+
+  def_options application: [
+                spec: allowed_applications(),
+                default: @default_application,
+                description: """
+                Output type (similar to compression amount). See https://opus-codec.org/docs/opus_api-1.3.1/group__opus__encoder.html#gaa89264fd93c9da70362a0c9b96b9ca88.
+                """
+              ],
+              input_caps: [
+                spec: Raw.t(),
+                type: :caps,
+                default: nil,
+                description: """
+                Input type - used to set input sample rate and channels
+                """
+              ]
+
+  def_input_pad :input, demand_unit: :bytes, caps: @supported_input
+  def_output_pad :output, caps: :any
+
+  @impl true
+  def handle_init(%__MODULE__{} = options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.merge(%{
+        native: nil,
+        queue: <<>>
+      })
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_start_of_stream(:input, _ctx, state) do
+    case state |> mk_native do
+      {:ok, native} ->
+        {:ok, %{state | native: native}}
+
+      {:error, reason} ->
+        {{:error, reason}, state}
+    end
+  end
+
+  @impl true
+  def handle_caps(:input, _caps, _ctx, _state) do
+    raise """
+    Changing input sample rate or channels is not supported
+    """
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _ctx, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, bufs, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, frame_size_in_bytes(state) * bufs}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
+    # holds a buffer of raw input that is not yet encoded
+    to_encode = state.queue <> data
+
+    case encode_buffer(to_encode, state, frame_size_in_bytes(state)) do
+      {:ok, {[], rest}} ->
+        # nothing was encoded
+        {{:ok, redemand: :output}, %{state | queue: rest}}
```

Comment:
@mat-hek you mentioned i don't have to redemand here - but what if the demand was in bytes that are not large enough to fill a frame? my understanding is that that's exactly when I would need to redemand.

Reply:
Oh ok, that's true. But then you always need to redemand, as it may happen that you generate some frames, but they are smaller in size than what was demanded.

Reply:
ah so if the downstream asks for 5 frames, I'm supposed to keep redemanding until I have at least 5 frames to deliver? I thought I redemand until I have _at least one_ frame. Still learning the framework :)